### PR TITLE
feature-multiple-columns-of-the-same-table

### PR DESCRIPTION
**Pull request by maksimryndin**
Thank you very much for your work.

I've tried on one column and results were outstanding.
In my case I have two columns in the same table to compress so decided to add such a feature to your library.

Also I've made a little fix on insert trigger (fields selection should be declared otherwise with several columns improper values are placed).

### CODE REVIEW

File: src/transparent.rs [hunk 1]
```diff
@@ -163,23 +165,61 @@ pub fn zstd_enable_transparent<'a>(ctx: &Context) -> anyhow::Result<ToSqlOutput<
...
+
+    let table_already_enabled = !already_compressed_columns.is_empty();
+
+    let dict_id_columns: Vec<String> = if table_already_enabled {
+        let query = format!(
+            r#"select "from"
+            from pragma_foreign_key_list('{}')
+            where "table" = '_zstd_dicts'"#,
+            &new_table_name
+        );
```
Comment (phiresky) [inline src/transparent.rs:195]: feels kinda weird to find these by foreign key, I'd probably have gotten them from already_compressed_columns? But I can't think of a specific reason why this is a bad idea, so I guess it's fine.

File: src/transparent.rs [hunk 2]
```diff
@@ -279,147 +318,328 @@ pub fn zstd_enable_transparent<'a>(ctx: &Context) -> anyhow::Result<ToSqlOutput<
...
+    let table_count: u32 = db
+        .query_row(
+            "select count(`type`) 
+            from sqlite_master 
+            where name = '_zstd_configs' and type = 'table'",
+            params![],
+            |r| r.get(0),
+        )
+        .with_context(|| format!("Could not get table information for _zstd_configs"))?;
+    if table_count == 0 {
```
Comment (phiresky) [inline src/transparent.rs:628]: i feel like the `check_table_exists` should be changed to only have the table_name argument, return a boolean and used here as well?

**Comment by phiresky** [comment]
Wow, this is great! I did write the code in a way that I thought should allow multiple columns in the future without major changes, and this patch is very tidy and simple!

Also, thank you for the fix with the insertion columns.



Note that if you use this library in "production", #1 needs to be fixed since that causes data loss if you open more than one DB in a process. (That might be the reason why some of the tests fail)

File: src/transparent.rs (hunk 2 above)
Comment (maksimryndin) [reply src/transparent.rs:628]: Sure) it is much better

File: src/transparent.rs (hunk 1 above)
Comment (maksimryndin) [reply src/transparent.rs:195]: Yes, it sounds good. But I believe quering db in this case is a more reliable way to get the current state of affairs:) in the sense of future features and refactoring

**Comment by maksimryndin** [comment]
Thank you very much for your review! Fixed in line with your advice. Also applied `cargo fmt` and `clippy`. Tests on my laptop pass in most cases but sometimes randomly some fail due to the issue #1 .

**Comment by phiresky** [comment]
Merged. I added this change which i think you missed: https://github.com/phiresky/sqlite-zstd/commit/2c394bf546f43295e6db61cc89023b17496d19ac

By the way, I'd be interested in what you're using this library for (e.g. what kind of data is in your DB, how many rows, how well does it compress) if you can say ;)

**Comment by maksimryndin** [comment]
> Merged. I added this change which i think you missed: [2c394bf](https://github.com/phiresky/sqlite-zstd/commit/2c394bf546f43295e6db61cc89023b17496d19ac)
> 
> By the way, I'd be interested in what you're using this library for (e.g. what kind of data is in your DB, how many rows, how well does it compress) if you can say ;)

Oh, thank a lot, I've really missed.
I can tell you the compress ratio  in a week as my data is still under preparation)) - very huge data.

So my task is to parse russian addresses (natural language understanding). So for the base data I use addresses data (relational kept in xml files) of Russian Fiscal Service (fias.nalog.ru). And during my parsing I collect address object hierarchy (for example, region -> district -> city -> street -> house) in two variant according to the russian territorial division (municipal/administrative - https://en.wikipedia.org/wiki/Political_divisions_of_Russia).

I need this hierarchy string further for the scoring during actual parsing of natural speech.

Hence two columns in the database with a lot of the same substrings (for example, hierarchy of all objects in the same region contains its region name).

I've tried your library for a tiny region of Adygea in Russia (db was 300 Mb before compression and 89 Mb after).
I expect my final db to be about 100 Gb and hope that after compression it will be much smaller. But also for the sake of size reduction I make null some the columns I don't need at parsing (the 300 Mb above do not include this step - it is a pure effect of sqlite-zstd).

**Pull request merged**